File: languages/go/entity/identity/credential_offer.go
package identity

import (
	"context"
	"errors"

	"github.com/bloock/bloock-sdk-go/v2/internal/bridge"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
	"github.com/bloock/bloock-sdk-go/v2/internal/config"
)

type CredentialOffer struct {
	Thid string
	Body CredentialOfferBody
	From string
	To   string
}

func NewCredentialOfferFromProto(s *proto.CredentialOffer) CredentialOffer {
	if s == nil {
		return CredentialOffer{}
	}
	return CredentialOffer{
		Thid: s.Thid,
		Body: NewCredentialOfferBodyFromProto(s.Body),
		From: s.XFrom,
		To:   s.XTo,
	}
}

func (c CredentialOffer) ToProto() *proto.CredentialOffer {
	return &proto.CredentialOffer{
		Thid:  c.Thid,
		Body:  c.Body.ToProto(),
		XFrom: c.From,
		XTo:   c.To,
	}
}

func NewCredentialOfferFromJson(json string) (CredentialOffer, error) {
	bridge := bridge.NewBloockBridge()
	res, err := bridge.Identity().CredentialOfferFromJson(context.Background(), &proto.CredentialOfferFromJsonRequest{
		ConfigData: config.NewConfigDataDefault(),
		Json:       json,
	})

	if err != nil {
		return CredentialOffer{}, err
	}

	if res.Error != nil {
		return CredentialOffer{}, errors.New(res.Error.Message)
	}

	return NewCredentialOfferFromProto(res.GetCredentialOffer()), nil
}

func (c CredentialOffer) ToJson() (string, error) {
	bridge := bridge.NewBloockBridge()
	res, err := bridge.Identity().CredentialOfferToJson(context.Background(), &proto.CredentialOfferToJsonRequest{
		ConfigData:      config.NewConfigDataDefault(),
		CredentialOffer: c.ToProto(),
	})

	if err != nil {
		return "", err
	}

	if res.Error != nil {
		return "", errors.New(res.Error.Message)
	}

	return res.GetJson(), nil
}

File: languages/js/src/entity/identity/credential_offer.ts
import { BloockBridge } from "../../bridge/bridge";
import {
  CredentialOfferToJsonRequest,
  CredentialOfferFromJsonRequest
} from "../../bridge/proto/identity";
import * as identityEntitiesProto from "../../bridge/proto/identity_entities";
import { NewConfigData } from "../../config/config";
import { CredentialOfferBody } from "./credential_offer_body";

export class CredentialOffer {
  thid: string;
  body: CredentialOfferBody;
  from: string;
  to: string;

  constructor(
    thid: string,
    body: CredentialOfferBody,
    from: string,
    to: string
  ) {
    this.thid = thid;
    this.body = body;
    this.from = from;
    this.to = to;
  }

  public toProto(): identityEntitiesProto.CredentialOffer {
    return identityEntitiesProto.CredentialOffer.fromPartial({
      thid: this.thid,
      body: this.body.toProto(),
      From: this.from,
      To: this.to
    });
  }

  static fromProto(r: identityEntitiesProto.CredentialOffer): CredentialOffer {
    return new CredentialOffer(
      r.thid,
      CredentialOfferBody.fromProto(r.body!),
      r.From,
      r.To
    );
  }

  public toJson(): Promise<string> {
    const bridge = new BloockBridge();

    const req = CredentialOfferToJsonRequest.fromPartial({
      configData: NewConfigData(undefined),
      credentialOffer: this.toProto()
    });

    return bridge
      .getIdentity()
      .CredentialOfferToJson(req)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.json;
      });
  }

  static fromJson(json: string): Promise<CredentialOffer> {
    const bridge = new BloockBridge();

    const req = CredentialOfferFromJsonRequest.fromPartial({
      configData: NewConfigData(undefined),
      json: json
    });

    return bridge
      .getIdentity()
      .CredentialOfferFromJson(req)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return CredentialOffer.fromProto(res.credentialOffer!);
      });
  }
}
